### PR TITLE
Issue #1: Validate developer properties

### PR DESCRIPTION
**Pull request by lmorchard**
Let's see how this looks to ya...

### CODE REVIEW

**Comment by lmorchard** [comment]
Oh wait, have to do some rebasing. Back in a sec.

**Comment by lmorchard** [comment]
Rebased, and simplified the recursive schema validation a bit

File: index.js [hunk 1]
```diff
@@ -25,6 +25,63 @@ var camelCase = function (word) {
...
+    parts.push(camelCase(part));
+  }
+  parts.push(camelCase(name));
+  return prefix + parts.join('');
+};
+
+var parseUrl = function (path) {
+  // TODO: Could replace this with a different implementation to make it
+  // node-agnostic and work in a browser. http://nodejs.org/api/url.html
+  return require('url').parse(path);
```
Comment (operatorjen) [inline index.js:40]: Yeah, this module is being used on the Marketplace backend, browser clientside verification and in devtools in WebIDE. So this parsing will have to be done with something that works in all cases.

File: index.js [hunk 2]
```diff
@@ -46,58 +103,84 @@ var Manifest = function () {
...
+      for (var k in schema.properties) {
+        if ('properties' in schema.properties[k] && k in subject) {
+          hasValidSchema(subject[k], schema.properties[k], parents.concat([k]));
+        }
+      }
+    }
+  };
+  
+  var hasMandatoryKeys = function (subject, schema, parents) {
+    if (!schema.required) { return; }
```
Comment (operatorjen) [inline index.js]: let's keep this on 3 lines (return on next line) for consistency

File: index.js [hunk 3]
```diff
@@ -46,58 +103,84 @@ var Manifest = function () {
...
-        errors['InvalidStringType' + camelCase(k)] = new Error('`' + k +
-               '` must be one of the following: ' + common.properties[k].oneOf.toString());
-      } else if (common.properties[k].anyOf) {
-        self.manifest[k].split(',').forEach(function (v) {
-          if (common.properties[k].anyOf.indexOf(v.trim()) === -1) {
-            errors['InvalidStringType' + camelCase(k)] = new Error('`' + k +
-               '` must be any of the following: ' + common.properties[k].anyOf.toString());
+  var hasValidStringItem = function (subject, schema, parents) {
+    for (var k in subject) {
+      if (!(k in schema.properties)) { continue; }
```
Comment (operatorjen) [inline index.js]: similar to above comment about 3 lines

File: index.js [hunk 4]
```diff
@@ -46,58 +103,84 @@ var Manifest = function () {
...
       }
     }
   };
 
-  var hasRequiredLength = function () {
-    for (var k in self.manifest) {
-      if (common.properties[k].minLength && self.manifest[k].toString().length < common.properties[k].minLength) {
+  var hasRequiredLength = function (subject, schema, parents) {
+    for (var k in subject) {
+      if (!(k in schema.properties)) { continue; }
```
Comment (operatorjen) [inline index.js]: same

File: test/test.index.js [hunk 5]
```diff
@@ -134,6 +134,51 @@ describe('validate', function () {
     should.not.exist(results.errors['InvalidPropertyLengthName']);
   });
 
+  it('should have an invalid developer name if empty', function () {
+    common.developer = {'name': ''};
```
Comment (operatorjen) [inline test/test.index.js]: same

File: test/test.index.js [hunk 6]
```diff
@@ -134,6 +134,51 @@ describe('validate', function () {
...
+
+    var results = m.validate(common);
+
+    results.errors['InvalidPropertyLengthDeveloperName'].toString().should.equal('Error: `name` must not be empty');
+  });
+
+  it('should have an invalid developer name if not string', function () {
+    common.developer = {
+      'name': { 'I have': 'no idea what I am doing' }
```
Comment (operatorjen) [inline test/test.index.js]: same

**Comment by operatorjen** [comment]
r+wc

**Comment by lmorchard** [comment]
Tweaked the block formatting, looking for an URL parser. (Or, might rework this to not need a full URL parser)

**Comment by operatorjen** [comment]
np! we can leave it as is. i'll merge it in and feel free to do that as a todo for the next iteration. this is more of an issue near the end of q3 rather than right now :)

**Comment by lmorchard** [comment]
Cool beans! Should probably even look into adding some browser-side testing on the travis-ci side along with that, wheee!

**Pull request merged**